File: program_stack/algorithms/strings/kmp/java/Kmp.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Kmp {
    /** LPS / failure function for pattern. */
    public static int[] buildLps(String pattern) {
        int m = pattern.length();
        int[] lps = new int[m];
        int len = 0, i = 1;
        while (i < m) {
            if (pattern.charAt(i) == pattern.charAt(len)) {
                lps[i++] = ++len;
            } else if (len > 0) {
                len = lps[len - 1];
            } else {
                lps[i++] = 0;
            }
        }
        return lps;
    }

    /** All starting indices of pattern in text. */
    public static List<Integer> kmpSearch(String text, String pattern) {
        List<Integer> hits = new ArrayList<>();
        if (pattern.isEmpty()) return hits;
        int[] lps = buildLps(pattern);
        int n = text.length(), m = pattern.length();
        int i = 0, q = 0;
        while (i < n) {
            if (text.charAt(i) == pattern.charAt(q)) {
                i++;
                q++;
                if (q == m) {
                    hits.add(i - m);
                    q = lps[q - 1];
                }
            } else if (q > 0) {
                q = lps[q - 1];
            } else {
                i++;
            }
        }
        return hits;
    }

    /** First match index, or -1. */
    public static int kmpFirst(String text, String pattern) {
        List<Integer> hits = kmpSearch(text, pattern);
        return hits.isEmpty() ? -1 : hits.get(0);
    }

    public static void main(String[] args) {
        List<Integer> hits = kmpSearch("ababcabab", "abab");
        if (!hits.equals(Arrays.asList(0, 5))) {
            System.err.println("unexpected hits: " + hits);
            System.exit(1);
        }
        if (kmpFirst("ababcabab", "abab") != 0) {
            System.err.println("first index failed");
            System.exit(1);
        }
        hits = kmpSearch("aaaa", "aa");
        if (!hits.equals(Arrays.asList(0, 1, 2))) {
            System.err.println("overlap hits: " + hits);
            System.exit(1);
        }
        if (kmpFirst("hello", "world") != -1) {
            System.err.println("miss should be -1");
            System.exit(1);
        }
        System.out.println("kmp: ok");
    }
}

File: program_stack/algorithms/strings/kmp/cpp/kmp.cpp
#include <iostream>
#include <string>
#include <vector>

std::vector<int> buildLps(const std::string& pattern) {
    int m = static_cast<int>(pattern.size());
    std::vector<int> lps(m, 0);
    int len = 0, i = 1;
    while (i < m) {
        if (pattern[i] == pattern[len]) {
            lps[i++] = ++len;
        } else if (len > 0) {
            len = lps[len - 1];
        } else {
            lps[i++] = 0;
        }
    }
    return lps;
}

std::vector<int> kmpSearch(const std::string& text, const std::string& pattern) {
    std::vector<int> hits;
    if (pattern.empty()) return hits;
    auto lps = buildLps(pattern);
    int n = static_cast<int>(text.size()), m = static_cast<int>(pattern.size());
    int i = 0, q = 0;
    while (i < n) {
        if (text[i] == pattern[q]) {
            ++i;
            ++q;
            if (q == m) {
                hits.push_back(i - m);
                q = lps[q - 1];
            }
        } else if (q > 0) {
            q = lps[q - 1];
        } else {
            ++i;
        }
    }
    return hits;
}

int kmpFirst(const std::string& text, const std::string& pattern) {
    auto hits = kmpSearch(text, pattern);
    return hits.empty() ? -1 : hits[0];
}

int main() {
    auto hits = kmpSearch("ababcabab", "abab");
    if (hits != std::vector<int>{0, 5}) {
        std::cerr << "unexpected hits\n";
        return 1;
    }
    if (kmpFirst("ababcabab", "abab") != 0) {
        std::cerr << "first index failed\n";
        return 1;
    }
    hits = kmpSearch("aaaa", "aa");
    if (hits != std::vector<int>{0, 1, 2}) {
        std::cerr << "overlap hits failed\n";
        return 1;
    }
    if (kmpFirst("hello", "world") != -1) {
        std::cerr << "miss should be -1\n";
        return 1;
    }
    std::cout << "kmp: ok\n";
    return 0;
}
